File: darkrodry/go/kata/game.go
package kata

type Game struct {
	currentRoll int
	rolls [21]int
}

func(g* Game) roll(pins int) {
	g.rolls[g.currentRoll] = pins
	g.currentRoll += 1
}

func(g* Game) score() int {
	score := 0
	frameIndex := 0
	for frame := 0; frame < 10; frame++ {
		if isStrike(g, frameIndex) {
			score += 10 + countStrikeBonus(g, frameIndex)
			frameIndex += 1
		} else if isSpare(g, frameIndex) {
			score += 10 + countSpareBonus(g, frameIndex)
			frameIndex += 2
		} else {
			score += countPointsInFrame(g, frameIndex)
			frameIndex += 2
		}
	}
	return score
}

func isStrike(g* Game, frameIndex int)  bool {
	return g.rolls[frameIndex] == 10
}

func isSpare(g* Game, frameIndex int) bool {
	return g.rolls[frameIndex] + g.rolls[frameIndex + 1] == 10
}

func countPointsInFrame(g* Game, frameIndex int) int {
	return g.rolls[frameIndex] + g.rolls[frameIndex + 1]
}

func countStrikeBonus(g* Game, frameIndex int) int {
	return g.rolls[frameIndex + 1] + g.rolls[frameIndex + 2]
}

func countSpareBonus(g* Game, frameIndex int) int {
	return g.rolls[frameIndex + 2]
}

File: darkrodry/ruby/kata/game.rb
class Game

  def initialize
    @current_roll = 0
    @rolls = Array.new(21) {0}
  end

  def roll(pins)
    @rolls[@current_roll] = pins
    @current_roll += 1
  end

  def score
    score = 0
    frame = 0
    10.times do
      if is_strike?(frame)
        score += 10 + count_strike_bonus(frame)
        frame += 1
      elsif is_spare?(frame)
        score += 10 + count_spare_bonus(frame)
        frame += 2
      else
        score += count_frame_points(frame)
        frame += 2
      end
    end
    return score
  end

  def is_strike?(frame)
    return @rolls[frame] == 10
  end

  def is_spare?(frame)
    return @rolls[frame] + @rolls[frame+1] == 10
  end

  def count_frame_points(frame)
    return @rolls[frame] + @rolls[frame+1]
  end

  def count_spare_bonus(frame)
    return @rolls[frame+2]
  end

  def count_strike_bonus(frame)
    return @rolls[frame+1] + @rolls[frame+2]
  end
end
